File: quarkdsl-web/lib/vm/lexer.ts
/**
 * QuarkDSL Lexer - DFA-based tokenization
 * Converts source code into a stream of tokens
 */

export enum TokenType {
  // Keywords
  Fn = "Fn",
  Let = "Let",
  Return = "Return",
  If = "If",
  Else = "Else",
  For = "For",
  In = "In",
  Map = "Map",

  // Annotations
  GpuAnnotation = "GpuAnnotation",
  QuantumAnnotation = "QuantumAnnotation",

  // Types
  Int = "Int",
  Float = "Float",
  Bool = "Bool",
  Qubit = "Qubit",
  Void = "Void",
  Tensor = "Tensor",
  QState = "QState",

  // Literals
  True = "True",
  False = "False",
  IntLiteral = "IntLiteral",
  FloatLiteral = "FloatLiteral",
  Identifier = "Identifier",

  // Operators
  Plus = "Plus",
  Minus = "Minus",
  Star = "Star",
  Slash = "Slash",
  Percent = "Percent",
  EqEq = "EqEq",
  Ne = "Ne",
  Lt = "Lt",
  Le = "Le",
  Gt = "Gt",
  Ge = "Ge",
  AndAnd = "AndAnd",
  OrOr = "OrOr",
  Bang = "Bang",
  Eq = "Eq",

  // Delimiters
  LParen = "LParen",
  RParen = "RParen",
  LBrace = "LBrace",
  RBrace = "RBrace",
  LBracket = "LBracket",
  RBracket = "RBracket",
  Comma = "Comma",
  Semicolon = "Semicolon",
  Colon = "Colon",
  Arrow = "Arrow",
  DotDot = "DotDot",

  // Special
  EOF = "EOF",
}

export interface Token {
  type: TokenType;
  value?: string | number;
  line: number;
  column: number;
}

const KEYWORDS: Record<string, TokenType> = {
  fn: TokenType.Fn,
  let: TokenType.Let,
  return: TokenType.Return,
  if: TokenType.If,
  else: TokenType.Else,
  for: TokenType.For,
  in: TokenType.In,
  map: TokenType.Map,
  int: TokenType.Int,
  float: TokenType.Float,
  bool: TokenType.Bool,
  qubit: TokenType.Qubit,
  void: TokenType.Void,
  tensor: TokenType.Tensor,
  qstate: TokenType.QState,
  true: TokenType.True,
  false: TokenType.False,
};

export class Lexer {
  private source: string;
  private pos: number = 0;
  private line: number = 1;
  private column: number = 1;

  constructor(source: string) {
    this.source = source;
  }

  private peek(offset: number = 0): string {
    return this.source[this.pos + offset] || "";
  }

  private advance(): string {
    const char = this.source[this.pos] || "";
    this.pos++;
    if (char === "\n") {
      this.line++;
      this.column = 1;
    } else {
      this.column++;
    }
    return char;
  }

  private skipWhitespace(): void {
    while (/\s/.test(this.peek())) {
      this.advance();
    }
  }

  private skipComment(): boolean {
    if (this.peek() === "/" && this.peek(1) === "/") {
      while (this.peek() && this.peek() !== "\n") {
        this.advance();
      }
      return true;
    }
    return false;
  }

  private skipWhitespaceAndComments(): void {
    while (true) {
      this.skipWhitespace();
      if (!this.skipComment()) {
        break;
      }
    }
  }

  private isAlpha(char: string): boolean {
    return /[a-zA-Z_]/.test(char);
  }

  private isAlphaNumeric(char: string): boolean {
    return /[a-zA-Z0-9_]/.test(char);
  }

  private isDigit(char: string): boolean {
    return /[0-9]/.test(char);
  }

  private makeToken(type: TokenType, value?: string | number): Token {
    return { type, value, line: this.line, column: this.column };
  }

  private readIdentifier(): Token {
    const startLine = this.line;
    const startCol = this.column;
    let value = "";

    while (this.isAlphaNumeric(this.peek())) {
      value += this.advance();
    }

    const type = KEYWORDS[value] || TokenType.Identifier;
    return {
      type,
      value: type === TokenType.Identifier ? value : undefined,
      line: startLine,
      column: startCol,
    };
  }

  private readNumber(): Token {
    const startLine = this.line;
    const startCol = this.column;
    let value = "";
    let isFloat = false;

    while (this.isDigit(this.peek())) {
      value += this.advance();
    }

    if (this.peek() === "." && this.isDigit(this.peek(1))) {
      isFloat = true;
      value += this.advance(); // consume '.'
      while (this.isDigit(this.peek())) {
        value += this.advance();
      }
    }

    return {
      type: isFloat ? TokenType.FloatLiteral : TokenType.IntLiteral,
      value: isFloat ? parseFloat(value) : parseInt(value, 10),
      line: startLine,
      column: startCol,
    };
  }

  tokenize(): Token[] {
    const tokens: Token[] = [];

    while (this.pos < this.source.length) {
      this.skipWhitespaceAndComments();

      if (this.pos >= this.source.length) break;

      const char = this.peek();
      const startLine = this.line;
      const startCol = this.column;

      // Annotations
      if (char === "@") {
        this.advance();
        let annotation = "";
        while (this.isAlpha(this.peek())) {
          annotation += this.advance();
        }
        if (annotation === "gpu") {
          tokens.push({
            type: TokenType.GpuAnnotation,
            line: startLine,
            column: startCol,
          });
        } else if (annotation === "quantum") {
          tokens.push({
            type: TokenType.QuantumAnnotation,
            line: startLine,
            column: startCol,
          });
        }
        continue;
      }

      // Identifiers and keywords
      if (this.isAlpha(char)) {
        tokens.push(this.readIdentifier());
        continue;
      }

      // Numbers
      if (this.isDigit(char)) {
        tokens.push(this.readNumber());
        continue;
      }

      // Two-character operators
      if (char === "-" && this.peek(1) === ">") {
        this.advance();
        this.advance();
        tokens.push({
          type: TokenType.Arrow,
          line: startLine,
          column: startCol,
        });
        continue;
      }
      if (char === "." && this.peek(1) === ".") {
        this.advance();
        this.advance();
        tokens.push({
          type: TokenType.DotDot,
          line: startLine,
          column: startCol,
        });
        continue;
      }
      if (char === "=" && this.peek(1) === "=") {
        this.advance();
        this.advance();
        tokens.push({
          type: TokenType.EqEq,
          line: startLine,
          column: startCol,
        });
        continue;
      }
      if (char === "!" && this.peek(1) === "=") {
        this.advance();
        this.advance();
        tokens.push({ type: TokenType.Ne, line: startLine, column: startCol });
        continue;
      }
      if (char === "<" && this.peek(1) === "=") {
        this.advance();
        this.advance();
        tokens.push({ type: TokenType.Le, line: startLine, column: startCol });
        continue;
      }
      if (char === ">" && this.peek(1) === "=") {
        this.advance();
        this.advance();
        tokens.push({ type: TokenType.Ge, line: startLine, column: startCol });
        continue;
      }
      if (char === "&" && this.peek(1) === "&") {
        this.advance();
        this.advance();
        tokens.push({
          type: TokenType.AndAnd,
          line: startLine,
          column: startCol,
        });
        continue;
      }
      if (char === "|" && this.peek(1) === "|") {
        this.advance();
        this.advance();
        tokens.push({
          type: TokenType.OrOr,
          line: startLine,
          column: startCol,
        });
        continue;
      }

      // Single-character tokens
      this.advance();
      switch (char) {
        case "+":
          tokens.push({
            type: TokenType.Plus,
            line: startLine,
            column: startCol,
          });
          break;
        case "-":
          tokens.push({
            type: TokenType.Minus,
            line: startLine,
            column: startCol,
          });
          break;
        case "*":
          tokens.push({
            type: TokenType.Star,
            line: startLine,
            column: startCol,
          });
          break;
        case "/":
          tokens.push({
            type: TokenType.Slash,
            line: startLine,
            column: startCol,
          });
          break;
        case "%":
          tokens.push({
            type: TokenType.Percent,
            line: startLine,
            column: startCol,
          });
          break;
        case "<":
          tokens.push({
            type: TokenType.Lt,
            line: startLine,
            column: startCol,
          });
          break;
        case ">":
          tokens.push({
            type: TokenType.Gt,
            line: startLine,
            column: startCol,
          });
          break;
        case "!":
          tokens.push({
            type: TokenType.Bang,
            line: startLine,
            column: startCol,
          });
          break;
        case "=":
          tokens.push({
            type: TokenType.Eq,
            line: startLine,
            column: startCol,
          });
          break;
        case "(":
          tokens.push({
            type: TokenType.LParen,
            line: startLine,
            column: startCol,
          });
          break;
        case ")":
          tokens.push({
            type: TokenType.RParen,
            line: startLine,
            column: startCol,
          });
          break;
        case "{":
          tokens.push({
            type: TokenType.LBrace,
            line: startLine,
            column: startCol,
          });
          break;
        case "}":
          tokens.push({
            type: TokenType.RBrace,
            line: startLine,
            column: startCol,
          });
          break;
        case "[":
          tokens.push({
            type: TokenType.LBracket,
            line: startLine,
            column: startCol,
          });
          break;
        case "]":
          tokens.push({
            type: TokenType.RBracket,
            line: startLine,
            column: startCol,
          });
          break;
        case ",":
          tokens.push({
            type: TokenType.Comma,
            line: startLine,
            column: startCol,
          });
          break;
        case ";":
          tokens.push({
            type: TokenType.Semicolon,
            line: startLine,
            column: startCol,
          });
          break;
        case ":":
          tokens.push({
            type: TokenType.Colon,
            line: startLine,
            column: startCol,
          });
          break;
        default:
          throw new Error(
            `Unexpected character '${char}' at line ${startLine}, column ${startCol}`
          );
      }
    }

    tokens.push({ type: TokenType.EOF, line: this.line, column: this.column });
    return tokens;
  }
}

File: src/frontend/lexer.rs
use logos::Logos;

// Lexical Analysis: DFA-based tokenization using Logos library
// Logos automatically generates DFA from regex patterns (Regex to DFA conversion)
// Thompson's Construction: Regex → ε-NFA → NFA → DFA (done internally by Logos)
// DFA Minimization: Logos optimizes the generated DFA

#[derive(Logos, Debug, Clone, PartialEq)]
#[logos(skip r"[ \t\n\f]+")]      // Skip whitespace (regex pattern)
#[logos(skip r"//[^\n]*")]        // Skip single-line comments (regex pattern)
pub enum Token {
    // Keywords (exact string matching in DFA)
    #[token("fn")]
    Fn,
    #[token("let")]
    Let,
    #[token("return")]
    Return,
    #[token("if")]
    If,
    #[token("else")]
    Else,
    #[token("for")]
    For,
    #[token("in")]
    In,
    #[token("map")]
    Map,

    // Annotations (domain-specific keywords)
    #[token("@gpu")]
    GpuAnnotation,
    #[token("@quantum")]
    QuantumAnnotation,

    // Types (keywords for type system)
    #[token("int")]
    Int,
    #[token("float")]
    Float,
    #[token("bool")]
    Bool,
    #[token("qubit")]
    Qubit,
    #[token("void")]
    Void,
    #[token("tensor")]
    Tensor,
    #[token("qstate")]
    QState,

    // Literals (constant values)
    #[token("true")]
    True,
    #[token("false")]
    False,

    // Regex to DFA: Integer literal pattern
    #[regex(r"[0-9]+", |lex| lex.slice().parse().ok())]
    IntLiteral(i64),

    // Regex to DFA: Float literal pattern (digits.digits)
    #[regex(r"[0-9]+\.[0-9]+", |lex| lex.slice().parse().ok())]
    FloatLiteral(f64),

    // Regex to DFA: Identifier pattern (letter/underscore followed by alphanumeric)
    // Maximal Munch: Longest match principle
    #[regex(r"[a-zA-Z_][a-zA-Z0-9_]*", |lex| lex.slice().to_string())]
    Identifier(String),

    // Operators
    #[token("+")]
    Plus,
    #[token("-")]
    Minus,
    #[token("*")]
    Star,
    #[token("/")]
    Slash,
    #[token("%")]
    Percent,

    #[token("==")]
    EqEq,
    #[token("!=")]
    Ne,
    #[token("<")]
    Lt,
    #[token("<=")]
    Le,
    #[token(">")]
    Gt,
    #[token(">=")]
    Ge,

    #[token("&&")]
    AndAnd,
    #[token("||")]
    OrOr,
    #[token("!")]
    Bang,

    #[token("=")]
    Eq,

    // Delimiters
    #[token("(")]
    LParen,
    #[token(")")]
    RParen,
    #[token("{")]
    LBrace,
    #[token("}")]
    RBrace,
    #[token("[")]
    LBracket,
    #[token("]")]
    RBracket,

    #[token(",")]
    Comma,
    #[token(";")]
    Semicolon,
    #[token(":")]
    Colon,
    #[token("->")]
    Arrow,
    #[token("..")]
    DotDot,
}

impl std::fmt::Display for Token {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            Token::Fn => write!(f, "fn"),
            Token::Let => write!(f, "let"),
            Token::Return => write!(f, "return"),
            Token::If => write!(f, "if"),
            Token::Else => write!(f, "else"),
            Token::For => write!(f, "for"),
            Token::In => write!(f, "in"),
            Token::Map => write!(f, "map"),
            Token::GpuAnnotation => write!(f, "@gpu"),
            Token::QuantumAnnotation => write!(f, "@quantum"),
            Token::Int => write!(f, "int"),
            Token::Float => write!(f, "float"),
            Token::Bool => write!(f, "bool"),
            Token::Qubit => write!(f, "qubit"),
            Token::Void => write!(f, "void"),
            Token::Tensor => write!(f, "tensor"),
            Token::QState => write!(f, "qstate"),
            Token::True => write!(f, "true"),
            Token::False => write!(f, "false"),
            Token::IntLiteral(n) => write!(f, "{}", n),
            Token::FloatLiteral(n) => write!(f, "{}", n),
            Token::Identifier(s) => write!(f, "{}", s),
            Token::Plus => write!(f, "+"),
            Token::Minus => write!(f, "-"),
            Token::Star => write!(f, "*"),
            Token::Slash => write!(f, "/"),
            Token::Percent => write!(f, "%"),
            Token::EqEq => write!(f, "=="),
            Token::Ne => write!(f, "!="),
            Token::Lt => write!(f, "<"),
            Token::Le => write!(f, "<="),
            Token::Gt => write!(f, ">"),
            Token::Ge => write!(f, ">="),
            Token::AndAnd => write!(f, "&&"),
            Token::OrOr => write!(f, "||"),
            Token::Bang => write!(f, "!"),
            Token::Eq => write!(f, "="),
            Token::LParen => write!(f, "("),
            Token::RParen => write!(f, ")"),
            Token::LBrace => write!(f, "{{"),
            Token::RBrace => write!(f, "}}"),
            Token::LBracket => write!(f, "["),
            Token::RBracket => write!(f, "]"),
            Token::Comma => write!(f, ","),
            Token::Semicolon => write!(f, ";"),
            Token::Colon => write!(f, ":"),
            Token::Arrow => write!(f, "->"),
            Token::DotDot => write!(f, ".."),
        }
    }
}

